doc/classes/MainLoop.xml

```diff
@@ -5,40 +5,53 @@
 	</brief_description>
 	<description>
 		[MainLoop] is the abstract base class for a Godot project's game loop. It is inherited by [SceneTree], which is the default game loop implementation used in Godot projects, though it is also possible to write and use one's own [MainLoop] subclass instead of the scene tree.
-		Upon the application start, a [MainLoop] implementation must be provided to the OS; otherwise, the application will exit. This happens automatically (and a [SceneTree] is created) unless a main [Script] is provided from the command line (with e.g. [code]godot -s my_loop.gd[/code], which should then be a [MainLoop] implementation.
+		Upon the application start, a [MainLoop] implementation must be provided to the OS; otherwise, the application will exit. This happens automatically (and a [SceneTree] is created) unless a [MainLoop] [Script] is provided from the command line (with e.g. [code]godot -s my_loop.gd[/code] or the "Main Loop Type" project setting is overwritten.
 		Here is an example script implementing a simple [MainLoop]:
-		[b]FIXME:[/b] No longer valid after DisplayServer split and Input refactoring.
 		[codeblock]
+		class_name CustomMainLoop
 		extends MainLoop
 
 		var time_elapsed = 0
-		var keys_typed = []
 		var quit = false
 
 		func _initialize():
 		    print("Initialized:")
 		    print("  Starting time: %s" % str(time_elapsed))
+		    _make_quit_action()
+
 
 		func _idle(delta):
 		    time_elapsed += delta
+
 		    # Return true to end the main loop.
-		    return quit
+		    return quit || Input.is_action_just_released("quit")
 
-		func _input_event(event):
-		    # Record keys.
-		    if event is InputEventKey and event.pressed and !event.echo:
-		        keys_typed.append(OS.get_keycode_string(event.keycode))
-		        # Quit on Escape press.
-		        if event.keycode == KEY_ESCAPE:
-		            quit = true
-		    # Quit on any mouse click.
-		    if event is InputEventMouseButton:
-		        quit = true
 
 		func _finalize():
 		    print("Finalized:")
 		    print("  End time: %s" % str(time_elapsed))
-		    print("  Keys typed: %s" % var2str(keys_typed))
+
+
+		# In your Main Loop you would have a viewport that you can use to gather input events
+		# You partly implement your quit logic in your viewport and then set quit to true to exit
```

Comment:
This example was a fully functioning example of a script you can execute, but it's no longer the case with these changes. This makes the example confusing IMO, `quit` is never set.

Reply:
Should I add a quit method?

Reply:
I don't know, what would the implementation be like?

The purpose of this example is to show how to make a very minimal MainLoop script that can handle basic things like counting time and receiving input to quit. It's just a demonstration of the use of MainLoop, it's not something useful per se and you don't have to stick to the original implementation if it's no longer meaningful in 4.0.

IMO, if input needs to go through the viewport, then showing how to handle input through the viewport would be useful.

Reply:
I rewrote that part using the Singleton directly.